core/schema/modulesource.go

```diff
@@ -1,1648 +1,2031 @@
 
-type parsedRefString struct {
-	modPath        string
-	modVersion     string
-	hasVersion     bool
-	kind           core.ModuleSourceKind
-	repoRoot       *vcs.RepoRoot
-	repoRootSubdir string
-	scheme         core.SchemeType
-	sshusername    string
+	return dagql.NewInstanceForCurrentID(ctx, s.dag, query, localSrc)
 }
 
-func (ref parsedRefString) String() string {
-	s := ref.modPath
-	if ref.scheme == core.SchemeSCPLike {
-		s = strings.Replace(s, "/", ":", 1)
-	}
-
-	if ref.sshusername != "" {
-		s = ref.sshusername + "@" + s
-	}
-	if ref.hasVersion {
-		s = s + "@" + ref.modVersion
+func callerHostFindUp(
+	ctx context.Context,
+	bk *buildkit.Client,
+	curDirPath string,
+	soughtName string,
+) (string, bool, error) {
+	found, err := callerHostFindUpAll(ctx, bk, curDirPath, map[string]struct{}{soughtName: {}})
+	if err != nil {
+		return "", false, err
 	}
-
-	s = ref.scheme.Prefix() + s
-	return s
-}
-
-// interface used for host interaction mocking
-type buildkitClient interface {
-	StatCallerHostPath(ctx context.Context, path string, followLinks bool) (*types.Stat, error)
+	p, ok := found[soughtName]
+	return p, ok, nil
 }
 
-// parseRefString parses a ref string into its components
-// New heuristic:
-// - stat folder to see if dir is present
-// - if not, try to isolate root of git repo from the ref
-// - if nothing worked, fallback as local ref, as before
-func parseRefString(ctx context.Context, bk buildkitClient, refString string) parsedRefString {
-	ctx, span := core.Tracer(ctx).Start(ctx, fmt.Sprintf("parseRefString: %s", refString), telemetry.Internal())
-	defer span.End()
-	localParsed := parsedRefString{
-		modPath: refString,
-		kind:    core.ModuleSourceKindLocal,
-		scheme:  core.NoScheme,
-	}
-
-	// if the refString is a relative path, we can short-circuit here as we
-	// don't really about the `stat` return as the refString will always
-	// be local in this case.
-	if strings.HasPrefix(refString, ".") {
-		return localParsed
-	}
+func callerHostFindUpAll(
+	ctx context.Context,
+	bk *buildkit.Client,
+	curDirPath string,
+	soughtNames map[string]struct{},
+) (map[string]string, error) {
+	found := make(map[string]string, len(soughtNames))
+	for {
+		for soughtName := range soughtNames {
+			stat, err := bk.StatCallerHostPath(ctx, filepath.Join(curDirPath, soughtName), true)
+			if err == nil {
+				delete(soughtNames, soughtName)
+				// NOTE: important that we use stat.Path here rather than curDirPath since the stat also
+				// does some normalization of paths when the client is using case-insensitive filesystems
+				found[soughtName] = filepath.Dir(stat.Path)
+				continue
+			}
+			// TODO: remove the strings.Contains check here (which aren't cross-platform),
+			// since we now set NotFound (since v0.11.2)
+			if status.Code(err) != codes.NotFound && !strings.Contains(err.Error(), "no such file or directory") {
+				return nil, fmt.Errorf("failed to lstat %s: %w", soughtName, err)
+			}
```

Comment:
If we bump the client <-> engine versions, we can remove this extra sadness as well :tada: